src/hotspot/share/gc/shared/collectedHeap.hpp

```diff
@@ -427,6 +430,8 @@ class CollectedHeap : public CHeapObj<mtGC> {
 
   virtual void initialize_serviceability() = 0;
 
+  virtual ObjectMarker* init_object_marker();
+
```

Comment:
Why does creation happen in collectedHeap.cpp, not objectMarker.cpp? And why expose this function?

Reply:
The intention here was to be able for a GC to provide an alternative implementation, especially in the context of Lilliput. See for example here: https://github.com/openjdk/lilliput/pull/32

Reply:
Sure, I get that :) I just wondered whether implementing this with, say, metaprogramming may be faster. E.g. make the controller a template and plugging in a Marker implementation that provides a "mark" function. Which then could be inlined. That way you don't pay for virtual calls at runtime. For a function that is invoked for every object this may be significant.

May also be done as a future RFE.

Reply:
> Sure, I get that :) I just wondered whether implementing this with, say, metaprogramming may be faster. E.g. make the controller a template and plugging in a Marker implementation that provides a "mark" function. Which then could be inlined. That way you don't pay for virtual calls at runtime. For a function that is invoked for every object this may be significant.
> 
> May also be done as a future RFE.

Right. Let me thinker with this idea for a bit.